src/groupeddataframe/splitapplycombine.jl

```diff
@@ -1167,30 +1135,55 @@ function _combine(f::AbstractVector{<:Pair},
                     if isnothing(agg2idx_map)
                         agg2idx_map = _agg2idx_map_helper(idx, idx_agg)
                     end
-                    res[i] = idx, res[i][2][agg2idx_map]
+                    res[i] = idx_agg, res[i][2][agg2idx_map]
                 elseif idx != res[i][1]
-                    throw(ArgumentError("all functions must return vectors of the same length"))
+                    if keeprows
+                        throw(ArgumentError("all functions must return vectors of " *
+                                            "the length equal to the group rows count " *
+                                            "in the source GroupedDataFrame"))
+                    else
+                        throw(ArgumentError("all functions must return vectors of the same length"))
+                    end
                 end
             end
         end
     end
+
+    # remember that here first field in res[i] is not useful - it is just needed
+    # to keep track how the column was generated
+    # a correct index is stored in idx variable
+
+    for (i, (col_idx, col)) in enumerate(res)
+        if keeprows && res[i][1] !== idx_keeprows # we need to reorder the column
+            newcol = similar(col)
+            # we can probably make it more efficient, but I leave it as an optimization for the future
+            for i in axes(col, 1)
+                newcol[gd.idx[i]] = col[i]
+            end
```

Comment:
```suggestion
            gd_idx = gd.idx
            @inbounds for j in eachindex(gd.idx, col)
                newcol[gd_idx[j]] = col[j]
            end
```

Reply:
I removed `@inbounds` for now. An `@inbounds` of this kind was a cause of `map` error not being caught for years (here we rely on the fact that `gd_idx` is a permutation - and it should be, but we make so many changes that I would prefer not to risk such bugs). This should be optimized in the future, that is why I have left a note above